### DISCUSSION TITLE
Right approach to make sense of JSON data

### DISCUSSION THREAD

**Discussion by rajyraman** [microsoft/semantic-kernel#7589] [Q&A] [answered]
I am pretty new to SK and LLMs in general. The use case I want to try is craft a prompt to search through a large JSON data that comes from API response. Initially I thought I can send it in the prompt itself as context, but then learnt that there are going to be too many tokens in that case and also it might exceed the token limit. So I am now exploring one of these two options:

1. Use function calling and do the retrieval and making sense of JSON data (from API) in the function itself. 
2. Use Semantic Memory and load the JSON into `VolatileMemoryStore`. Then craft the prompt to search this data.

What is the better approach to follow, or is there any other different approach?

**Comment by dmytrostruk** [comment]
Hi @rajyraman !
I'm wondering, if you need to perform a search through a large JSON data that comes from API response, do you need to send a prompt at all? I think you can convert your JSON to embeddings and then perform a vector search with your query to find specific data in this JSON? In this case, there is no need in function calling, and using `VolatileMemoryStore` or other memory connector to retrieve related data should be enough. Let us know if this could cover your case. Thank you!

**Reply by rajyraman** [reply]
@dmytrostruk - I was thinking function calling only because I am not sure what kind of questions about the data the user might ask, which means I can even load the data from server and query it only _after_ the user asks the question in which case I can do all the manipulations in the function/multiple functions itself. 

Unlike documents, JSON makes sense only it its entirety so I am not sure if you can write a prompt that will return the JSON summarised by a column after filtering it by a different column etc.

**Reply by rajyraman** [reply]
@dmytrostruk - I was looking at this sample and wondering if this is the new way to do it.

https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/Agents/OpenAIAssistant_FileManipulation.cs

**Reply by rajyraman** [reply]
@dmytrostruk - According to [this](https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0026-file-service.md) I cannot using file service with AOI.
![image](https://github.com/user-attachments/assets/83ef7bf6-152c-4f51-b164-e3d121e332db)

**Reply by dmytrostruk** [reply]
> I was thinking function calling only because I am not sure what kind of questions about the data the user might ask, which means I can even load the data from server and query it only after the user asks the question in which case I can do all the manipulations in the function/multiple functions itself.

@rajyraman Thanks for provided information. It would be very helpful if you could provide some examples how the JSON will look like and what kind of questions you expect to handle. This will help to understand how to achieve this scenario. Thanks!

**Reply by rajyraman** [reply]
@dmytrostruk - I tried a very basic example which is no where near the complexity of the response JSON and it does not work the way I thought it would.
![image](https://github.com/user-attachments/assets/9e17075a-508c-4703-91ef-bf60395c3488)

**Reply by dmytrostruk** [reply]
@rajyraman In case your expected result is just `#f00`, there are two ways:
1. Using just vector search - but in this case you will have to chunk your JSON by each property or choose some strategy for chunking based on the nature of your JSON, so your search results will be more accurate. In example you shared above, you saved all the information about colors as one record, so that's why you received such response. On the other hand, you also saved another record with link and as you can see, your response is still related to colors, so vector search works correctly, but you need to define your chunking strategy to get better results.
2. Using RAG with vector search + function calling - in this case, you still perform vector search to get your JSON information from vector store, but you can receive better response by asking AI to analyze this JSON part and return specific result. Here are some examples with similar scenario - https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/RAG/WithPlugins.cs

**Reply by rajyraman** [reply]
@dmytrostruk - I really like the experience that I see in https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/Agents/OpenAIAssistant_FileManipulation.cs. Because I don't have to deal with any of the embedding, retrieval bit. Just feed the file and you can query with prompt. The is what I see as ideal experience.  But seems like it is not possible with Azure OAI. 

The problem with RAG + Function calling for me is that the information I need would be impossible without looking at the entire JSON. If I inject the entire JSON into the prompt I do get the right answer, but obviously this approach would not work on a JSON with 10K lines.

![image](https://github.com/user-attachments/assets/8985fd43-9b7a-45b0-b7eb-666da97d7da0)

Even if I chunk by each object in the colours array, to answer the question "Give me the list of colours" I need to feed all the 7 colour properties on the prompt (after RAG), which on a JSON with 10K lines most likely will exceed the token limit.

**Reply by dmytrostruk** [reply]
> I really like the experience that I see in https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/Agents/OpenAIAssistant_FileManipulation.cs. Because I don't have to deal with any of the embedding, retrieval bit. Just feed the file and you can query with prompt. The is what I see as ideal experience. But seems like it is not possible with Azure OAI.

@rajyraman If you are interested in Assistant API and file manipulation, it's possible to use it with Azure OpenAI. The example you shared above use constructor for communication with OpenAI, but it also has constructor overload which accepts `endpoint` parameter, which you can use to pass Azure OpenAI Endpoint URL. You could try this approach and let us know if it covers your scenario. Thanks!

**Reply by rajyraman** [reply]
@dmytrostruk  - It does not work with the Azure Open AI endpoint.

![image](https://github.com/user-attachments/assets/83275dc1-f34e-48a9-bab6-466027cc34cc)

This is exactly as per https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0026-file-service.md#context-and-problem-statement which says "Azure Open AI does not currently support the OpenAI file service API."

Please clarify which constructor accepts in AOI endpoint URL.

**Reply by dmytrostruk** [reply]
> Please clarify which constructor accepts in AOI endpoint URL.

Sorry, I was talking about `OpenAIFileService`, which is used in this example. In example, the service is initialized with OpenAI key, but there is another constructor that accepts Azure OpenAI settings:
https://github.com/microsoft/semantic-kernel/blob/3b0d086a50193e6a5a3809c3ef74538dad778362/dotnet/src/Connectors/Connectors.OpenAI/Files/OpenAIFileService.cs#L42-L53

> This is exactly as per https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0026-file-service.md#context-and-problem-statement which says "Azure Open AI does not currently support the OpenAI file service API."

@crickman Could you please provide an information about ADR? Is it up-to-date?

**Reply by rajyraman** [reply]
@dmytrostruk - According to this https://github.com/microsoft/semantic-kernel/blob/feature_agents_openai/dotnet/docs/OPENAI-CONNECTOR-MIGRATION.md#4-openai-file-service-deprecation lot of things have still not been merged into main yet. Once https://github.com/microsoft/semantic-kernel/pull/7126 is merged I can probably try it or is there any other way?

**Reply by dmytrostruk** [reply]
> According to this https://github.com/microsoft/semantic-kernel/blob/feature_agents_openai/dotnet/docs/OPENAI-CONNECTOR-MIGRATION.md#4-openai-file-service-deprecation lot of things have still not been merged into main yet. Once https://github.com/microsoft/semantic-kernel/pull/7126 is merged I can probably try it or is there any other way?

@rajyraman This part of work is a migration to new OpenAI SDK. After it's merged, it's possible that code syntax or namespaces will change a little bit, but in terms of functionality everything should work pretty much the same. If you are interested in Assistants API, I think it's possible to achieve your scenario with current functionality in `main` branch using Azure OpenAI. @crickman will provide more information if needed.

**Comment by Cobra86** [comment]
I'm using Azure AI Search with SK. It's working perfectly. The only issue is that Azure AI Search/SK only supports one index.

**Comment by crickman** [comment, marked as answer]
@rajyraman - Our implementation of `OpenAIFileService` does appear to struggle with Azure endpoint for certain interactions; although, it was designed to support targeting both _OpenAI_ and _Azure Open AI_ services.

The current version of SK is based on V1 SDK's and we have a **major update** _just around the corner_ where we've migrated to the V2 SDK's.  This update impacts the memory connectors and the assistant-agents.  As part of this, we've made `obsolete` the `OpenAIFileService` in favor of direclty using the `FileClient` from the V2 SDK.  This sample has been updated to include this change in a feature-branch and is expected to be merged to main in the _very near future_!!!

In the meantime, I've re-written this sample to use the V1 `AssistantsClient` and confirmed functionality against our team's Azure deployments.

```c#
// Copyright (c) Microsoft. All rights reserved.
using System.Text;
using Azure;
using Azure.AI.OpenAI.Assistants;
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Agents;
using Microsoft.SemanticKernel.Agents.OpenAI;
using Microsoft.SemanticKernel.ChatCompletion;
using Resources;

namespace Agents;

/// <summary>
/// Demonstrate using code-interpreter to manipulate and generate csv files with <see cref="OpenAIAssistantAgent"/> .
/// </summary>
public class OpenAIAssistant_FileManipulation(ITestOutputHelper output) : BaseTest(output)
{
    /// <summary>
    /// Target OpenAI services.
    /// </summary>
    protected override bool ForceOpenAI => false;

    [Fact]
    public async Task AnalyzeCSVFileUsingOpenAIAssistantAgentAsync()
    {
        AssistantsClient client = new(new Uri(TestConfiguration.AzureOpenAI.Endpoint), new AzureKeyCredential(TestConfiguration.AzureOpenAI.ApiKey));
        await using Stream stream = EmbeddedResource.ReadStream("sales.csv")!;
        OpenAIFile fileInfo = await client.UploadFileAsync(stream, OpenAIFilePurpose.Assistants);

        Console.WriteLine(this.ApiKey);

        // Define the agent
        OpenAIAssistantAgent agent =
            await OpenAIAssistantAgent.CreateAsync(
                kernel: new(),
                config: new(this.ApiKey, this.Endpoint),
                new()
                {
                    EnableCodeInterpreter = true, // Enable code-interpreter
                    ModelId = this.Model,
                    FileIds = [fileInfo.Id] // Associate uploaded file
                });

        // Create a chat for agent interaction.
        AgentGroupChat chat = new();

        // Respond to user input
        try
        {
            await InvokeAgentAsync("Which segment had the most sales?");
            await InvokeAgentAsync("List the top 5 countries that generated the most profit.");
            await InvokeAgentAsync("Create a tab delimited file report of profit by each country per month.");
        }
        finally
        {
            await agent.DeleteAsync();
            await client.DeleteFileAsync(fileInfo.Id);
        }

        // Local function to invoke agent and display the conversation messages.
        async Task InvokeAgentAsync(string input)
        {
            chat.AddChatMessage(new ChatMessageContent(AuthorRole.User, input));

            Console.WriteLine($"# {AuthorRole.User}: '{input}'");

            await foreach (ChatMessageContent content in chat.InvokeAsync(agent))
            {
                Console.WriteLine($"# {content.Role} - {content.AuthorName ?? "*"}: '{content.Content}'");

                foreach (AnnotationContent annotation in content.Items.OfType<AnnotationContent>())
                {
                    Console.WriteLine($"\n* '{annotation.Quote}' => {annotation.FileId}");
                    BinaryData fileContent = await client.GetFileContentAsync(annotation.FileId!);
                    Console.WriteLine(Encoding.Default.GetString(fileContent.ToArray()));
                }
            }
        }
    }
}
```

**Reply by rajyraman** [reply]
@crickman - Is there a specific version I should use? I am getting this exception.

![image](https://github.com/user-attachments/assets/ceb7715e-d9ab-431d-b931-b54640b730b3)
![image](https://github.com/user-attachments/assets/632c065e-8b32-439f-b204-ecc86d68fb5d)

**Reply by rajyraman** [reply]
@crickman - Seems like I can use the JSON file to do search using Code Interpreter. Now I need to understand why the code is not working.
![image](https://github.com/user-attachments/assets/64eb9aa3-a309-4b49-a959-d8a26e6d4a2c)
